### PR TITLE
KYLIN-4342 Build Global Dict by MR/Hive New Version

### PR DESCRIPTION
**Pull request by wangxiaojing123**
issue link: https://issues.apache.org/jira/browse/KYLIN-4342

### CODE REVIEW

**Comment by hit-lacus** [comment]
Hi, @wangxiaojing123 , I see this PR contains 9 commits, which contains all code change which is necessary for [Build Global Dict by MR/Hive New Version](https://issues.apache.org/jira/browse/KYLIN-4342).

So, if I make no mistake, maybe you could close all other previous PRs, like this one https://github.com/apache/kylin/pull/1096, because the current PR is the superset of that pr.

File: engine-mr/src/main/java/org/apache/kylin/engine/mr/steps/BuildGlobalHiveDictTotalBuildMapper.java [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+            buffer= new BufferedReader(new InputStreamReader(in, "UTF-8")) ;
+            String line = null;
+            while ((line = buffer.readLine()) != null) {
+                stat.append(line);
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }finally {
+            buffer.close();
+            in.close();
```
Comment (hit-lacus) [inline engine-mr/src/main/java/org/apache/kylin/engine/mr/steps/BuildGlobalHiveDictTotalBuildMapper.java]: Maybe checkIfNotNull before close it for `in` and `buffer` is better, see https://lgtm.com/projects/g/apache/kylin/rev/pr-41895d7e0679c810b18a84e71eb8221f677676e0.

**Comment by nichunen** [comment]
Agree with @hit-lacus

File: engine-mr/src/main/java/org/apache/kylin/engine/mr/steps/BuildGlobalHiveDictTotalBuildMapper.java (hunk 1 above)
Comment (wangxiaojing123) [reply engine-mr/src/main/java/org/apache/kylin/engine/mr/steps/BuildGlobalHiveDictTotalBuildMapper.java]: OK，thanks.

File: engine-mr/src/main/java/org/apache/kylin/engine/mr/steps/BuildGlobalHiveDicPartBuildReducer.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    @Override
+    public void doReduce(Text key, Iterable<LongWritable> values, Context context)
+            throws IOException, InterruptedException {
+        count++;
+        byte[] keyBytes = Bytes.copy(key.getBytes(), 1, key.getLength() - 1);
+        if(count==1){
+            colIndex = key.getBytes()[0];//col index
+            colName = dicCols[colIndex];
+        }
+        logAFewRows(key.toString());
```
Comment (hit-lacus) [inline engine-mr/src/main/java/org/apache/kylin/engine/mr/steps/BuildGlobalHiveDicPartBuildReducer.java]: I am not totally sure, but I suspected this `toString` maybe cause some gc overhead beacuse it create a lot of `String` objects?

Comment (wangxiaojing123) [reply engine-mr/src/main/java/org/apache/kylin/engine/mr/steps/BuildGlobalHiveDicPartBuildReducer.java]: OK, I will fix it.

**Comment by hit-lacus** [comment]
Close this because it is introduced in https://github.com/apache/kylin/pull/1207 .

**Pull request closed without merging**